nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -0,0 +1,142 @@
+          }
+        }
+        index++;
+      }
+      // Generics check for nested type parameters
+      for (Type typeArgument : typeArguments) {
+        if (typeArgument.getTypeArguments().length() > 0) {
+          checkInstantiatedType(typeArgument, state, tree, analysis, config);
+        }
+      }
+    }
+  }
+
+  /** Generics checks for parameterized typed trees * */
+  public static void checkInstantiationForParameterizedTypedTree(
+      ParameterizedTypeTree tree, VisitorState state, NullAway analysis, Config config) {
+    if (!config.isJSpecifyMode()) {
+      throw new IllegalStateException("should only check type instantiations in JSpecify mode");
+    }
+    List<? extends Tree> typeArguments = tree.getTypeArguments();
+    HashSet<Integer> nullableTypeArguments = new HashSet<Integer>();
+    JCTree.JCTypeApply baseTypeTree = (JCTree.JCTypeApply) tree;
+    Type t = baseTypeTree.type;
+    Type.ClassType classType = (Type.ClassType) t;
+    // If none of the arguments are @Nullable annotated MetaData is empty
+    if (classType.tsym.getMetadata() != null) {
+      List<Attribute.TypeCompound> baseTypeAttributes =
+          classType.tsym.getMetadata().getTypeAttributes();
+      // Store the arguments in the base type that have @Nullable annotation in the set
+      for (int i = 0; i < baseTypeAttributes.size(); i++) {
+        // position - index of the parameters in the list of base type attributes that have
+        // @Nullable annotation
+        if (baseTypeAttributes.get(i).toString().equals("@org.jspecify.nullness.Nullable")) {
+          nullableTypeArguments.add(baseTypeAttributes.get(i).position.parameter_index);
+        }
+      }
+    }
+    for (int i = 0; i < typeArguments.size(); i++) {
+      if (typeArguments.get(i).getClass().equals(JCTree.JCTypeApply.class)) {
+        ParameterizedTypeTree parameterizedTypeTreeForTypeArgument =
+            (ParameterizedTypeTree) typeArguments.get(i);
+        Type argumentType = ASTHelpers.getType(parameterizedTypeTreeForTypeArgument);
+        if (argumentType != null
+            && argumentType.getTypeArguments() != null
+            && argumentType.getTypeArguments().length() > 0) { // Nested generics
+          checkInstantiationForParameterizedTypedTree(
+              parameterizedTypeTreeForTypeArgument, state, analysis, config);
+        }
+      }
+
+      boolean hasNullableAnnotation = false;
+      if (typeArguments.get(i).getClass().equals(JCTree.JCAnnotatedType.class)) {
+        JCTree.JCAnnotatedType annotatedType = (JCTree.JCAnnotatedType) typeArguments.get(i);
+        for (JCTree.JCAnnotation annotation : annotatedType.getAnnotations()) {
+          Attribute.Compound attribute = annotation.attribute;
+          hasNullableAnnotation =
+              hasNullableAnnotation
+                  || attribute.toString().equals("@org.jspecify.nullness.Nullable");
+          if (hasNullableAnnotation) {
+            break;
+          }
+        }
+      }
```

Comment:
So then is this the code that checks for `@Nullable` annotations in the `ParameterizedTypeTree` itself (not on upper bounds)?